ci/filter-tasks.js

```diff
@@ -0,0 +1,156 @@
+// File to filter out tasks that don't need to be built
+var fs = require('fs');
+var path = require('path');
+var restClient = require('typed-rest-client/RestClient');
+var httpHandler = require('typed-rest-client/Handlers');
+var run = require('./ci-util').run;
+
+var makeOptionsPath = path.join(__dirname, '..', 'make-options.json');
+var makeOptions = JSON.parse(fs.readFileSync(makeOptionsPath).toString());
+
+var getTasksToBuildFromVersioning = async function() {
+    // Returns a list of tasks that have different version numbers than their current published version.
+    var token = process.env['SYSTEM_ACCESSTOKEN'];
+    var packageInfo;
+    try {
+        var handler = new httpHandler.PersonalAccessTokenCredentialHandler(token);
+        var client = new restClient.RestClient('Tasks CI', '', [handler]);
+        packageInfo = await client.get('https://mseng.feeds.visualstudio.com/_apis/packaging/Feeds/Codex-deps/Packages?packageNameQuery=Mseng.MS.TF.DistributedTask.Tasks');
+        if (packageInfo.statusCode != 200) {
+            return makeOptions.tasks;
+        }
+    }
+    catch (err) {
+        // If unable to reach rest client, build everything.
+        return makeOptions.tasks;
+    }
+    var packageMap = {};
+    packageInfo.result.value.forEach(package => {
+        if (package.name && package.versions) {
+            var packageName = package.name.slice('Mseng.MS.TF.DistributedTask.Tasks.'.length);
+            var packageVersion = package.versions[0].version;
+            package.versions.forEach(versionInfo => {
+                if (versionInfo.isLatest) {
+                    packageVersion = versionInfo.version;
+                }
+            });
+            packageMap[packageName] = packageVersion;
+        }
+    });
+
+    return makeOptions.tasks.filter(function (taskName) {
+        if (taskName in packageMap) {
+            var packageVersion = packageMap[taskName].split('.');
+            var packageMajorVersion = packageVersion[0];
+            var packageMinorVersion = packageVersion[1];
+            var packagePatchVersion = packageVersion[2];
+
+            var taskJsonPath = path.join(__dirname, '..', 'Tasks' , taskName, 'task.json')
+            var taskJson = JSON.parse(fs.readFileSync(taskJsonPath).toString());
+            var version = taskJson.version;
+            if (packageMajorVersion != version.Major || packageMinorVersion != version.Minor || packagePatchVersion != version.Patch) {
+                // If versions are not the same, want to build this package
```

Comment:
You can use semver module to compare versions. We want to be careful here. It's not just that they are unequal, it's that the local version is greater than the packaging version. We can also log that they are equal(easier to debug later) and report a warning if it's less than.

Reply:
Logging equality and warning if it is lower makes sense to me, can you think of a case where ending up with a lower package number would be valid though? Maybe it makes more sense to fail the build on this?